File: _old_csharp/addons/terrabrush/KeybindManager.cs
#if TOOLS
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Godot;
using Godot.Collections;

namespace TerraBrush;

public partial class KeybindManager : RefCounted {
	// EditorSettings Group: terrabrush
	private const string SettingsGroup = "terrabrush";

	#region InputMap Actions and associated Default Key Events
	private readonly InputEventKey ToolPieKey = new InputEventKey() { Keycode = Key.V };
    private readonly InputEventKey BrushPieKey = new InputEventKey() { Keycode = Key.B };
    private readonly InputEventKey ToolContentKey = new InputEventKey() { Keycode = Key.N };
    private readonly InputEventKey BrushSizeKey = new InputEventKey() { Keycode = Key.G };
    private readonly InputEventKey BrushStrengthKey = new InputEventKey() { Keycode = Key.H };
    private readonly InputEventKey EscapeKey = new InputEventKey() { Keycode = Key.Escape };
    private readonly InputEventKey ToggleAutoAddZonesKey = new InputEventKey() { Keycode = Key.K };
    private readonly InputEventKey LockXAxis = new InputEventKey() { Keycode = Key.X };
    private readonly InputEventKey LockZAxis = new InputEventKey() { Keycode = Key.Z };
    #endregion

    #region Public String Names

    public static class StringNames
    {
	    public static readonly StringName ToolPie = new StringName("terrabrush_tool_pie");
	    public static readonly StringName BrushPie = new StringName("terrabrush_brush_pie");
	    public static readonly StringName ToolContentPie = new StringName("terrabrush_tool_content_pie");
	    public static readonly StringName BrushSizeSelector = new StringName("terrabrush_brush_size_selector");
	    public static readonly StringName BrushStrengthSelector = new StringName("terrabrush_brush_strength_selector");
	    public static readonly StringName EscapeSelector = new StringName("terrabrush_brush_escape_selector");
	    public static readonly StringName ToggleAutoAddZones = new StringName("terrabrush_toggle_auto_add_zones");
	    public static readonly StringName LockXAxis = new StringName("terrabrush_lock_x_axis");
	    public static readonly StringName LockZAxis = new StringName("terrabrush_lock_z_axis");
    }
    #endregion

	private readonly List<StringName> _actionNames;
	private readonly System.Collections.Generic.Dictionary<StringName, InputEventKey> _defaultKeys;

    // Quick access to Action Manes and Default Keys to register with the Godot Editor
    public List<StringName> ActionNames => _actionNames;
    public System.Collections.Generic.Dictionary<StringName, InputEventKey> DefaultKeys => _defaultKeys;

    public KeybindManager() {
	    _actionNames = new List<StringName>() {
		    StringNames.ToolPie,
			StringNames.BrushPie,
			StringNames.ToolContentPie,
		    StringNames.BrushSizeSelector,
			StringNames.BrushStrengthSelector,
			StringNames.EscapeSelector,
			StringNames.ToggleAutoAddZones,
			StringNames.LockXAxis,
			StringNames.LockZAxis,
	    };
	    _defaultKeys = new System.Collections.Generic.Dictionary<StringName, InputEventKey>() {
		    { StringNames.ToolPie, ToolPieKey },
		    { StringNames.BrushPie, BrushPieKey },
		    { StringNames.ToolContentPie, ToolContentKey},
		    { StringNames.BrushSizeSelector, BrushSizeKey },
		    { StringNames.BrushStrengthSelector, BrushStrengthKey },
		    { StringNames.EscapeSelector, EscapeKey },
		    { StringNames.ToggleAutoAddZones, ToggleAutoAddZonesKey },
		    { StringNames.LockXAxis, LockXAxis },
		    { StringNames.LockZAxis, LockZAxis },
	    };
    }

    public static Array<InputEvent> GetBinding(StringName actionName) => InputMap.Singleton.ActionGetEvents(actionName);

    public static string DescribeKey(InputEventKey key) {
	    var desc = new StringBuilder();
	    if (key.CtrlPressed) desc.Append("Ctrl +");
	    if (key.AltPressed) desc.Append("Alt +");
	    desc.Append(key.Keycode.ToString());
	    return desc.ToString();
    }

    /// <summary>
    /// Register's the Action Names with the Godot Editor InputMap.
    /// </summary>
    public void RegisterInputMap() {
	    var im = InputMap.Singleton;
	    foreach (var action in ActionNames) {
			if (!im.HasAction(action)) {
				im.AddAction(action); // Register the Action, as we don't have the Action.
		    }
	    }
    }

    /// <summary>
    /// Load Keybinds from Godot Editor's EditorSettings File, for association with the InputMap.Actions.
    /// </summary>
	public void LoadEditorSettings() {
		var settings = EditorInterface.Singleton.GetEditorSettings();
		var im = InputMap.Singleton;
		var shortcuts = settings.GetSetting("shortcuts").AsGodotArray<Godot.Collections.Dictionary>();
		foreach (var action in ActionNames) {
			var binding = $"{SettingsGroup}/{action}";
			var keyEvents = shortcuts.FirstOrDefault(x => x["name"].AsString() == binding);
			if (keyEvents != null) {
				// Register Saved event in EditorSettings.
				var keys = keyEvents["shortcuts"].AsGodotArray<InputEventKey>();
				foreach (var key in keys) {
					im.ActionAddEvent(action, key);
				}
			} else {
				// Register Default Event
				im.ActionAddEvent(action, DefaultKeys[action]);
			}
		}
	}

    /// <summary>
    /// Saves the Keybinds to Godot Editor's EditorSettings file, for associated InputMap.Actions.
    /// </summary>
	public void SaveEditorSettings() {
		var im = InputMap.Singleton;
		var settings = EditorInterface.Singleton.GetEditorSettings();
		var shortcuts = settings.GetSetting("shortcuts").AsGodotArray<Godot.Collections.Dictionary>();
		foreach (var action in ActionNames) {
			var binding = $"{SettingsGroup}/{action}";
			var keyEvents = shortcuts.FirstOrDefault(x => x["name"].AsString() == binding);
			if (keyEvents == null) {
				var keybind = new Godot.Collections.Dictionary<string, Variant>();
				keybind["name"] = binding;
				keybind["shortcuts"] = im.ActionGetEvents(action);
				shortcuts.Add((Godot.Collections.Dictionary)keybind);
			}
			else {
				keyEvents["shortcuts"] = im.ActionGetEvents(action);
			}
		}

		settings.SetSetting("shortcuts", shortcuts);
	}

	public void UpdateKeybind(StringName action, InputEventKey key) {
		var im = InputMap.Singleton;
		im.ActionEraseEvents(action);
		im.ActionAddEvent(action, key);
	}

	public void ResetKeybind(StringName action) {
		if (!ActionNames.Contains(action)) return;
		var im = InputMap.Singleton;
		im.ActionEraseEvents(action);
		im.ActionAddEvent(action, DefaultKeys[action]);
	}

	public string DescribeKey(StringName action) {
		var key = InputMap.Singleton.ActionGetEvents(action);
		return DescribeKey((InputEventKey)key[0]);
	}
}
#endif

File: src/misc/keybind_manager.cpp
#include "keybind_manager.h"

#include <godot_cpp/classes/input_map.hpp>
#include <godot_cpp/classes/editor_interface.hpp>
#include <godot_cpp/classes/editor_settings.hpp>
#include <godot_cpp/classes/os.hpp>

using namespace godot;

void KeybindManager::_bind_methods() {

}

TypedArray<Ref<InputEvent>> KeybindManager::getBinding(StringName actionName) {
    return InputMap::get_singleton()->action_get_events(actionName);
}

String KeybindManager::describeInputKey(Ref<InputEventKey> key) {
    String desc = "";
    if (key->is_ctrl_pressed()) desc += "Ctrl + ";
    if (key->is_alt_pressed()) desc += "Alt + ";
    desc += OS::get_singleton()->get_keycode_string(key->get_keycode());
    return desc;
}

const StringName KeybindManager::StringNames::ToolPie() { return StringName("terrabrush_tool_pie"); }
const StringName KeybindManager::StringNames::BrushPie() { return StringName("terrabrush_brush_pie"); }
const StringName KeybindManager::StringNames::ToolContentPie() { return StringName("terrabrush_tool_content_pie"); }
const StringName KeybindManager::StringNames::BrushSizeSelector() { return StringName("terrabrush_brush_size_selector"); }
const StringName KeybindManager::StringNames::BrushStrengthSelector() { return StringName("terrabrush_brush_strength_selector"); }
const StringName KeybindManager::StringNames::EscapeSelector() { return StringName("terrabrush_brush_escape_selector"); }
const StringName KeybindManager::StringNames::ToggleAutoAddZones() { return StringName("terrabrush_toggle_auto_add_zones"); }
const StringName KeybindManager::StringNames::LockXAxis() { return StringName("terrabrush_lock_x_axis"); }
const StringName KeybindManager::StringNames::LockZAxis() { return StringName("terrabrush_lock_z_axis"); }

KeybindManager::KeybindManager() {
    _toolPieKey = Ref<InputEventKey>(memnew(InputEventKey));
    _toolPieKey->set_keycode(Key::KEY_V);

    _brushPieKey = Ref<InputEventKey>(memnew(InputEventKey));
    _brushPieKey->set_keycode(Key::KEY_B);

    _toolContentKey = Ref<InputEventKey>(memnew(InputEventKey));
    _toolContentKey->set_keycode(Key::KEY_N);

    _brushSizeKey = Ref<InputEventKey>(memnew(InputEventKey));
    _brushSizeKey->set_keycode(Key::KEY_G);

    _brushStrengthKey = Ref<InputEventKey>(memnew(InputEventKey));
    _brushStrengthKey->set_keycode(Key::KEY_H);

    _escapeKey = Ref<InputEventKey>(memnew(InputEventKey));
    _escapeKey->set_keycode(Key::KEY_ESCAPE);

    _toggleAutoAddZonesKey = Ref<InputEventKey>(memnew(InputEventKey));
    _toggleAutoAddZonesKey->set_keycode(Key::KEY_K);

    _lockXAxis = Ref<InputEventKey>(memnew(InputEventKey));
    _lockXAxis->set_keycode(Key::KEY_X);

    _lockZAxis = Ref<InputEventKey>(memnew(InputEventKey));
    _lockZAxis->set_keycode(Key::KEY_Z);

    _actionNames = TypedArray<StringName>({
        StringNames::ToolPie(),
        StringNames::BrushPie(),
        StringNames::ToolContentPie(),
        StringNames::BrushSizeSelector(),
        StringNames::BrushStrengthSelector(),
        StringNames::EscapeSelector(),
        StringNames::ToggleAutoAddZones(),
        StringNames::LockXAxis(),
        StringNames::LockZAxis(),
    });
    _defaultKeys = TypedDictionary<StringName, InputEventKey>();
    _defaultKeys[StringNames::ToolPie()] = _toolPieKey;
    _defaultKeys[StringNames::BrushPie()] = _brushPieKey;
    _defaultKeys[StringNames::ToolContentPie()] = _toolContentKey;
    _defaultKeys[StringNames::BrushSizeSelector()] = _brushSizeKey;
    _defaultKeys[StringNames::BrushStrengthSelector()] = _brushStrengthKey;
    _defaultKeys[StringNames::EscapeSelector()] = _escapeKey;
    _defaultKeys[StringNames::ToggleAutoAddZones()] = _toggleAutoAddZonesKey;
    _defaultKeys[StringNames::LockXAxis()] = _lockXAxis;
    _defaultKeys[StringNames::LockZAxis()] = _lockZAxis;
}

KeybindManager::~KeybindManager() {}

TypedArray<StringName> KeybindManager::get_actionNames() const {
    return _actionNames;
}

TypedDictionary<StringName, Ref<InputEventKey>> KeybindManager::get_defaultKeys() const {
    return _defaultKeys;
}

/// <summary>
/// Register's the Action Names with the Godot Editor InputMap.
/// </summary>
void KeybindManager::registerInputMap(bool registerDefaultKey) {
    InputMap *im = InputMap::get_singleton();
    for (StringName action : _actionNames) {
        if (!im->has_action(action)) {
            im->add_action(action); // Register the Action, as we don't have the Action.
            if (registerDefaultKey) {
                im->action_add_event(action, _defaultKeys[action]);
            }
        }
    }
}

/// <summary>
/// Load Keybinds from Godot Editor's EditorSettings File, for association with the InputMap.Actions.
/// </summary>
void KeybindManager::loadEditorSettings() {
    Ref<EditorSettings> settings = EditorInterface::get_singleton()->get_editor_settings();
    InputMap *im = InputMap::get_singleton();
    TypedArray<Dictionary> shortcuts = settings->get_setting("shortcuts");
    for (StringName action : _actionNames) {
        String binding = String(SettingsGroup) + "/" + action;

        Dictionary keyEvents;

        for (Dictionary shortcut : shortcuts) {
            if (shortcut["name"] == binding) {
                keyEvents = shortcut;
                break;
            }
        }

        if (!keyEvents.is_empty()) {
            // Register Saved event in EditorSettings.
            TypedArray<Ref<InputEventKey>> keys = keyEvents["shortcuts"];
            for (Ref<InputEventKey> key : keys) {
                im->action_add_event(action, key);
            }
        }
        else {
            // Register Default Event
            im->action_add_event(action, _defaultKeys[action]);
        }
    }
}

void KeybindManager::saveEditorSettings() {
    Ref<EditorSettings> settings = EditorInterface::get_singleton()->get_editor_settings();
    InputMap *im = InputMap::get_singleton();
    TypedArray<Dictionary> shortcuts = settings->get_setting("shortcuts");
    for (StringName action : _actionNames) {
        String binding = String(SettingsGroup) + "/" + action;

        Dictionary keyEvents;

        for (Dictionary shortcut : shortcuts) {
            if (shortcut["name"] == binding) {
                keyEvents = shortcut;
                break;
            }
        }

        if (keyEvents.is_empty()) {
            TypedDictionary<String, Variant> keybind = TypedDictionary<String, Variant>();
            keybind["name"] = binding;
            keybind["shortcuts"] = im->action_get_events(action);
            shortcuts.append(keybind);
        }
        else {
            keyEvents["shortcuts"] = im->action_get_events(action);
        }
    }

    settings->set_setting("shortcuts", shortcuts);
}

void KeybindManager::updateKeybind(StringName action, Ref<InputEventKey> key) {
    InputMap *im = InputMap::get_singleton();
    im->action_erase_events(action);
    im->action_add_event(action, key);
}

void KeybindManager::resetKeybind(StringName action) {
    if (!_actionNames.has(action)) return;
    InputMap *im = InputMap::get_singleton();
    im->action_erase_events(action);
    im->action_add_event(action, _defaultKeys[action]);
}

String KeybindManager::describeKey(StringName action) {
    TypedArray<Ref<InputEvent>> key = InputMap::get_singleton()->action_get_events(action);
    return describeInputKey(key[0]);
}
